packages/flutter/lib/src/material/menu_bar.dart

```diff
@@ -0,0 +1,3274 @@
+    // Setting the menuStyle value will automatically check for changes and
+    // notify listeners.
+    if (value == null) {
+      _entry.menuStyle = _entry.menuStyle?.copyWithout(alignment: true);
+      return;
+    }
+    _entry.menuStyle = _entry.menuStyle?.copyWith(alignment: value) ?? MenuStyle(alignment: value);
+  }
+
+  /// Sets the alignment offset of the menu relative to the alignment position
+  /// specified by [alignment] on the controlling widget that the menu is
+  /// attached to.
+  ///
+  /// The offset is a directional offset starting at the alignment point, so
+  /// increasingly larger positive values of [Offset.dx] will offset from
+  /// 'begin' moving towards 'end' from the alignment point specified by
+  /// [alignment] on the controlling widget. This will either offset the menu
+  /// to the right (for [TextDirection.ltr]) or left (for [TextDirection.rtl]),
+  /// depending on the ambient [Directionality].
+  Offset get alignmentOffset => _entry.alignmentOffset;
+  set alignmentOffset(Offset value) {
+    assert(ChangeNotifier.debugAssertNotDisposed(this));
+    // Setting the _entry.alignmentOffset value will automatically check for
+    // changes and notify listeners.
+    _entry.alignmentOffset = value;
+  }
+
+  /// Sets the [MenuStyle] to use for configuring the menu.
+  ///
+  /// Setting this value will change the visual presentation of the menu to
+  /// match the given theme. Setting it to null will return the menu to default
+  /// values derived from the ambient [MenuTheme].
+  ///
+  /// If the menu is already open, then the theme for the menu will be updated.
+  MenuStyle? get menuStyle => _entry.menuStyle;
+  set menuStyle(MenuStyle? value) {
+    assert(ChangeNotifier.debugAssertNotDisposed(this));
+    // Setting the _entry value will automatically check for changes and notify listeners.
+    _entry.menuStyle = value;
+  }
+
+  /// Whether or not the associated menu is currently open.
+  bool get isOpen => _entry.isOpen;
+
+  /// Open the menu.
+  ///
+  /// Call this from the controlling widget when the menu should open up.
+  void open() {
+    assert(ChangeNotifier.debugAssertNotDisposed(this));
+    _entry.open();
+  }
+
+  /// Close the menu.
+  ///
+  /// Call this when the menu should be closed. Has no effect if the menu is already closed.
+  void close() {
+    assert(ChangeNotifier.debugAssertNotDisposed(this));
+    _entry.close();
+  }
+
+  /// Dispose of the menu.
+  ///
+  /// Must be called when the menu is no longer needed, typically when the
+  /// controlling widget is disposed.
+  @override
+  void dispose() {
+    assert(ChangeNotifier.debugAssertNotDisposed(this));
+    _entry.removeListener(notifyListeners);
+    _entry.dispose();
+    super.dispose();
+  }
+}
+
+/// Creates a new cascading menu given the focus node for the controlling
+/// widget.
+///
+/// Calling `createMaterialMenu` creates a new cascading menu controlled by
```

Comment:
How does this work if the menu is to be shown when the user clicks right somewhere on the app? In that case the default origin of the menu should be aligned with the origin of the mouse event,

Reply:
Currently, that isn't possible.  I need a `BuildContext` in order to copy the themes into the overlay, and a single point doesn't give me that context, so I required a `GlobalKey` (I used to require a `FocusNode`, which also has a `BuildContext`, but I decided that was too restrictive).  I could make a different function that takes a `Rect` and a `BuildContext`, but the problem then is that the build context isn't guaranteed to be valid next frame, so I'm back to taking a `GlobalKey` which will update itself. I could add an optional `Rect` to use instead of the rect from the `GlobalKey`'s context, however, so maybe that's the solution, but you'd still need to supply a GlobalKey that lets me know what context I should copy the themes from. Once we have the `OverlayPortal` in, that could simplify this a lot.

Reply:
Maybe the controller's open method should have an optional  parameter that's the global coordinates of the menu's origin (subject to adjustment of course).

Reply:
There already is an `alignmentOffset` member on `MenuEntry` that you can update, and it will update the location of the menu automatically. It is in relative coordinates from the alignment position, so if you specify `Alignment.topLeft` as the alignment, then it is in local coordinates of the rect that the `GlobalKey`'s context describes.

I could add an optional `Offset` to the "open" that would take a global coordinate, and then translate that to a local coordinate internally and set the `alignmentOffset` from that so that it would be internally consistent.

Reply:
Actually, resolving the correct local `alignmentOffset` is actually really hard at the point where we're opening the menu, so I think instead I'll just have an attribute that is normally null, but if you set it, overrides the `alignment`/`alignmentOffset` settings and just puts it at that global location.

Reply:
That seems reasonable to me. A context or background menu that pops up under the mouse on right click is a really common desktop UI idiom, so an example or a feature of an existing example that demo'd as much would be helpful.